src/tree/updater_quantile_hist.cc

```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
       this->EvaluateSplit(cright, gmat, hist_, *p_fmat, *p_tree);
       time_evaluate_split += dmlc::GetTime() - tstart;
 
-      qexpand_->push(ExpandEntry(cleft, p_tree->GetDepth(cleft),
+      qexpand_loss_guided_->push(ExpandEntry(cleft, p_tree->GetDepth(cleft),
                                  snode_[cleft].best.loss_chg,
                                  timestamp++));
-      qexpand_->push(ExpandEntry(cright, p_tree->GetDepth(cright),
+      qexpand_loss_guided_->push(ExpandEntry(cright, p_tree->GetDepth(cright),
                                  snode_[cright].best.loss_chg,
                                  timestamp++));
 
       ++num_leaves;  // give two and take one, as parent is no longer a leaf
     }
   }
+}
+
+void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+                                    const GHistIndexBlockMatrix& gmatb,
+                                    const ColumnMatrix& column_matrix,
+                                    HostDeviceVector<GradientPair>* gpair,
+                                    DMatrix* p_fmat,
+                                    RegTree* p_tree) {
+  double gstart = dmlc::GetTime();
+
+  const std::vector<GradientPair>& gpair_h = gpair->ConstHostVector();
+
+  spliteval_->Reset();
+
+  tstart = dmlc::GetTime();
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  time_init_data = dmlc::GetTime() - tstart;
+
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1)
+      << "tree_method=hist does not support multiple roots at this moment";
+  if (param_.grow_policy == TrainParam::kLossGuide) {
+     ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+    while (!qexpand_loss_guided_->empty()) {
```

Comment:
I actually feel this loop is not necessary, @hcho3 ?

Reply:
@CodingCat This loop exists so that the queue `qexpand_loss_guided_` gets emptied properly at the end. Note that some choice of split evaluators may result in early termination of the algorithm, with a bunch of leaf nodes left in the queue. This loop ensures that the nodes actually get labelled as leaf.

Reply:
but for loss guide, we always jump out from the while loop https://github.com/dmlc/xgboost/blob/master/src/tree/updater_quantile_hist.cc#L140, do you mean in other future cases?

Reply:
You're actually right; the loop is currently not needed. See comments 
https://github.com/dmlc/xgboost/blob/9b917cda4ff61f0512cdba1c53bc8e55258a5d4a/src/tree/updater_quantile_hist.cc#L197-L199

Reply:
Let's just remove this loop, along with the 3-line comments?